core/field.js

```diff
@@ -885,6 +888,10 @@ Blockly.Field.prototype.doClassValidation_ = function(newValue) {
  * @protected
  */
 Blockly.Field.prototype.doValueUpdate_ = function(newValue) {
+  // If the value is valid that is the value we want to be displayed. So set
+  // the display value to the new value.
+  this.displayValue_ = newValue;
```

Comment:
As shown in the diagram I posted below, this is solving for fields that want to show invalid values.

> case where the validator strips half the text out
That case involves a changed value.

Most fields (i.e. all fields except the text input) should accept changed values, since they are valid.

Ignoring changed values can be built ontop of this system, as I show in this demo.